### PR TITLE
[core] Delete disconnected node view in ray syncer when connection is broken.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

The current ray syncer doesn't take care of disconnection very well. If one raylet is disconnected due to some reason, the ray syncer won't clear its local view and will send the dead node info to any newly joined node.

This won't introduce any correctness bugs because the raylet will just reject the offending message.
And this won't introduce too big performance impact since only the newly added node will receive these mesages.

This PR cleaned up its local view table when the node is disconnect. It'll get the new snapshot when it rejoin if the disconnection is due to network.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/ray_syncer/ray_syncer.cc [hunk 1]
```diff
@@ -173,13 +177,11 @@ RaySyncer::RaySyncer(instrumented_io_context &io_context,
...
 RaySyncer::~RaySyncer() {
   *stopped_ = true;
-  io_context_.dispatch(
-      [reactors = sync_reactors_]() {
-        for (auto [_, reactor] : reactors) {
-          reactor->Disconnect();
-        }
-      },
-      "");
+  boost::asio::dispatch(io_context_.get_executor(), [reactors = sync_reactors_]() {
```
Comment (rkooo567) [inline src/ray/common/ray_syncer/ray_syncer.cc:180]: what's the difference between this vs io_context_.dispatch(?

File: src/ray/common/ray_syncer/ray_syncer.cc [hunk 2]
```diff
@@ -231,19 +230,21 @@ void RaySyncer::Connect(const std::string &node_id,
...
+                node_state_->RemoveNode(node_id);
               }
             },
             /* stub */ std::move(stub));
         Connect(reactor);
         reactor->StartCall();
-      },
-      "");
+      }))
+      .get();
```
Comment (rkooo567) [inline src/ray/common/ray_syncer/ray_syncer.cc:242]: Does it mean this will be a blocking call?

**Review by rkooo567 — COMMENTED** [review]
Q: why do we change io_context_.dispath to boost::asio::dispatch in this PR?

File: src/ray/common/ray_syncer/ray_syncer.cc [hunk 3]
```diff
@@ -197,23 +199,20 @@ std::shared_ptr<const RaySyncMessage> RaySyncer::GetSyncMessage(
...
-      "");
-  return promise.get_future().get();
+  auto task = std::packaged_task<std::vector<std::string>()>([&]() {
+    std::vector<std::string> nodes;
+    for (auto [node_id, _] : sync_reactors_) {
+      nodes.push_back(node_id);
+    }
+    return nodes;
+  });
+  return boost::asio::dispatch(io_context_.get_executor(), std::move(task)).get();
```
Comment (iycheng) [inline src/ray/common/ray_syncer/ray_syncer.cc:209]: @rkooo567 it's for better way to write blocking call. 
I think `io_context_.dispatch` is the old style  and asio now has executor.

I don't think there is big difference. Just playing with new things so we'll be able to get better idea of how to write asio code (we highly depend on asio and I feel we are not using it in a correct way :( )

File: src/ray/common/ray_syncer/ray_syncer.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/common/ray_syncer/ray_syncer.cc:242]: Yes, it'll. It returns a future.

**Comment by iycheng** [comment]
@rkooo567  no big different with our current usage of asio. But I feel it's a better way to write asio code and I'm trying to play around with this to get better idea.

Here with the new way, we can make it return a future and we just get to make it block.

**Review by rkooo567 — COMMENTED** [review]
One additional questiosn regarding the change in unblocking -> blocking

File: src/ray/common/ray_syncer/ray_syncer.cc (hunk 3 above)
Comment (iycheng) [reply src/ray/common/ray_syncer/ray_syncer.cc:209]: I think future actually should be returned so the caller can decide what to do.

A better way is to accept a completion token and the caller can decide what to do with that. I'd say we can adopt this pattern in gRPC refactoring work in the future.

**Comment by iycheng** [comment]
@rkooo567 my original motivation is only to make things simpler and easier to understand. I can revert it back if you have strong opinion on this.
I think it won't cause a deadlock since it's dispatch.

**Review by jjyao — APPROVED** [review]

File: src/ray/common/ray_syncer/ray_syncer.cc (hunk 2 above)
Comment (jjyao) [reply src/ray/common/ray_syncer/ray_syncer.cc:242]: Will it affect perf like blocking the thread for too long?

**Comment by iycheng** [comment]
synced offline with @rkooo567 . merge it.

**Pull request merged**